Herencia

EmanuelCl opened the pull request:
hola! estoy un poco trabado en como encarar la clase registro del ejercicio de la ultima clase... cree la clase padre Vehículo y los hijos auto, camión y moto, pero nose como hacer para leerlos y cargarlos los 3 en el mismo registro.

francisco-corti commented:
> hola! estoy un poco trabado en como encarar la clase registro del ejercicio de la ultima clase... cree la clase padre Vehículo y los hijos auto, camión y moto, pero nose como hacer para leerlos y cargarlos los 3 en el mismo registro.

Buenas! Necesitas algo para diferenciar que tipo de vehiculo es cada registro. Una de las posibles soluciones sería que en el txt guardes el tipo de vehiculo en el primer campo de cada linea, por ej:

`auto;ABC123;Ford;Fiesta;2012
`

Y al momento de instanciar los objetos, evaluar su tipo con un switch o varios if. ej:

```
switch(propiedadesVehiculo[0]){
  case "auto":
  new Auto(propiedadesVehiculo[1], ...);
  break;
  ...
```

EmanuelCl commented:
Estaria bien planteado asi?

import * as RLS from 'readline-sync';
import * as FS from 'fs';
import Vehiculo from "./Vehiculo";
import Auto from './Auto';
import Moto from './Moto';
import Camion from './Camion';
export default class RegistroVehiculos {
    private vehiculos: Vehiculo[];
    private autos: Auto []
    private motos: Moto []
    private camion: Camion []

    public constructor (){
        this.vehiculos = [];
        this.autos = [];
        this.motos = [];
        this.camion = [];
    }

    public addVehiculo():void{
        let tipo:string = RLS.question("Ingrese el tipo de vehiculo: ").toUpperCase();
        let patente:string = RLS.question('Ingrese la Patente: ');
        let marca:string = RLS.question('Ingrese la marca: ');
        let modelo:string = RLS.question('Ingrese el modelo: ');
        let año: number = RLS.questionInt('Ingrese el año: ');
        
        switch(tipo){
            case "auto":{
                let combustible:string = RLS.question("Ingrese el tipo de Combustible: ")
                this.autos.push(new Auto(tipo,patente,marca,modelo,año,combustible));
                break;
            }
            case "moto":{
                let cilindrada:number = RLS.questionInt("Ingrese la Cilindrada: ");
                this.motos.push(new Moto(tipo,patente,marca,modelo,año,cilindrada));
                break;
            }
            case "camion":{
                let cargaKg:number = RLS.questionInt("Ingrese la capacidad de Carga: ");
                this.camion.push(new Camion(tipo,patente,marca,modelo,año,cargaKg));
                break;
            }
        }
    }

    public findVehiculo(patente:string): number{
        for (let i = 0; i < this.vehiculos.length; i++) {
            if(patente == this.vehiculos[i].getPatente()){
                return i;
            }
        }
        return -1;
    }

    public deleteVehiculo(patente:string): void{
        let posicion:number = this.findVehiculo(patente);
        if (posicion != -1) {
            this.vehiculos.splice(posicion,1);
        }
    }

    public updateVehiculo(vehiculoViejo: string): void{
        
        let posicion= this.findVehiculo(vehiculoViejo);
        if (posicion != -1) {
            this.vehiculos[posicion] = new Vehiculo(tipo,patente,marca,modelo,año)
        }else{
            console.log("el vehiculo no existe");
        }
    }

    public mostrarVehiculo(): void{
        console.log(this.vehiculos);
    }

    public cargarVehiculo(rutaArchivo:string,separador:string):void{
        let vehiculos:string[] = (FS.readFileSync(rutaArchivo,'utf8')).split(separador);
        let propiedadVehiculo:string[] = [];
        vehiculos.forEach(vehiculoString => {
            propiedadVehiculo = vehiculoString.split(";")
        
            switch(propiedadVehiculo[0]){
                case "auto":{
                    this.autos.push(new Auto(propiedadVehiculo[0],propiedadVehiculo[1],propiedadVehiculo[2],propiedadVehiculo[3],parseInt(propiedadVehiculo[4]),propiedadVehiculo[5]));
                    break;
                }
                case "moto":{
                    this.motos.push(new Moto(propiedadVehiculo[0],propiedadVehiculo[1],propiedadVehiculo[2],propiedadVehiculo[3],parseInt(propiedadVehiculo[4]),parseInt(propiedadVehiculo[5])));
                    break;
                }
                case "camion":{
                    this.camion.push(new Camion(propiedadVehiculo[0],propiedadVehiculo[1],propiedadVehiculo[2],propiedadVehiculo[3],parseInt(propiedadVehiculo[4]),parseInt(propiedadVehiculo[5])));
                    break;
                }
            }
        });
    }
}

francisco-corti commented:
Está bien planteado Ema.

> ```
> export default class RegistroVehiculos {
> private vehiculos: Vehiculo[];
> private autos: Auto []
> private motos: Moto []
> private camion: Camion []
> 
>
> public constructor (){
>     this.vehiculos = [];
>     this.autos = [];
>     this.motos = [];
>     this.camion = [];
> }
> ```

Acá te alcanzaría con solo tener el array de **Vehiculo** como variable de estado. Ya que las instancias de clase que la extienden (auto, moto, camion), implicitamente tambien son de tipo Vehiculo, por lo cual no deberia haber problemas al pushear todos los objetos a ese mismo array.

mauri52 commented on Programacion Orientada a Objetos/Clase 27/Automotor/RegistroVehiculos.ts:
```diff
@@ -0,0 +1,100 @@
+  
+    public constructor (){
+        this.vehiculos = [];
+    
```
Como bien dice Fran, esto tiene que ser un array de Vehiculos que es la clase padre y la que tiene la mayoria de los atributos. La herencia precisamente se utiliza para aprovechar eso. Aunque en algun lado tenes que indicar cual es la clase de cada objeto, es decir vos creas instancias de Auto, Moto y Camion, pero los almacenas como Vehiculos (ya que todos los autos son vehiculos, al igual que las motos y los camiones), pero cuando los queres manipular por separado debes saber de que clase o tipo es cada uno, ya que no le podrias aplicar por ejemplo getCilindrada a un Camion

mauri52 commented on Programacion Orientada a Objetos/Clase 27/Automotor/Vehiculo.ts:
```diff
@@ -0,0 +1,35 @@
+export default class Vehiculo{
+    private tipo:string;
```
ese atributo contesta mi comentario anterior

mauri52 reviewed (approved):
Muy bien Ema.
